src/util.jl

```diff
@@ -162,49 +162,101 @@ function solve_triangular_system(d::OrderedDict)
     OrderedDict{Symbol,Real}(zip(nms, data))
 end
 
+type TriangularSystemException <: Exception
+    missing
+end
 
 """
 Solves triangular system specified by incidence dictionary.
 
-Returns a Vector of keys specifying the order in which `dd` should be evaluated
-
 ```
-system = Dict(0=>[1,2], 1=>Int[], 2=>[1] )
+system = Dict(0=>[1,2], 1=>[], 2=>[1] )
 solve_dependency(system)
 ```
 
 or
 
 ```
-system = Dict(:x=>[:y,:z], :y=>Symbol[], :z=>[:y])
+system = Dict(:x=>[:y,:z], :y=>[], :z=>[:y] )
 solve_dependency(system)
 ```
+
+Optionally, one can add specify which subset of variables to solve for.
+Unrequired variables will be ommited in the solution.
+
 """
-function solve_dependency(dd::Dict{T,Set{T}}) where T # is hashable
-    solved = T[]
-    deps = deepcopy(dd)
-    p = length(deps)
-    it = 0
-    while length(deps) > 0 && it <= p
-        it += 1
-        for (k, dep) in deps
-            if length(dep) == 0
-                push!(solved, k)
-                pop!(deps, k)
-                for (l, ldeps) in deps
-                    if k in ldeps
-                        pop!(ldeps, k)
-                    end
+function solve_dependencies(deps::Associative{T,Set{T}}, needed=nothing) where T
+    solution = []
+    if needed == nothing
```

Comment:
Probably not a first order concern, but needed is not type stable because of this if/else block.

Maybe we should name the argument to the function `_needed` and keep the left hand side variable defined in this block equal to `needed`. Or we can change the lefthand side variable from this block to a different name.

Reply:
Sure